File: admin/src/types/logistics.ts
/**
 * Types for Logistics and Operations Module
 * Phase 4: Picking, Packing, Shipping, and Route Optimization
 */

// ============================================================================
// PICKING & PACKING
// ============================================================================

export type PickingType = 'single_order' | 'batch' | 'wave';
export type PickingStatus = 'pending' | 'in_progress' | 'completed' | 'cancelled';
export type PackingStatus = 'pending' | 'in_progress' | 'completed' | 'problem';

export interface PickingList {
  id: string;
  picking_number: string;
  workspace_id: number;

  // Type
  type: PickingType;

  // Orders
  sale_ids: string[];

  // Items
  items: PickingItem[];

  // Optimized route
  picking_route: string[]; // Order of locations

  // Status
  status: PickingStatus;

  // Team
  assigned_to: string;
  assigned_to_name?: string;
  started_at?: Date;
  completed_at?: Date;

  // Metrics
  estimated_time: number; // minutes
  actual_time?: number;

  // Priority
  priority: 'low' | 'normal' | 'high' | 'urgent';

  created_at: Date;
  updated_at: Date;
}

export interface PickingItem {
  id: string;
  product_id: string;
  product_name: string;
  product_sku: string;

  quantity_ordered: number;
  quantity_picked: number;

  location: string; // Warehouse location
  batch_id?: string;
  serial_numbers?: string[];

  picked: boolean;
  picker_id?: string;
  picked_at?: Date;

  // Issues
  has_issue: boolean;
  issue_type?: 'out_of_stock' | 'damaged' | 'wrong_product' | 'other';
  issue_notes?: string;
}

export interface PackingStation {
  id: string;
  name: string;
  warehouse_id: string;
  code: string;

  // Configuration
  packer_id?: string;
  packer_name?: string;
  is_active: boolean;

  // Available materials
  available_boxes: BoxType[];

  // Metrics
  orders_packed_today: number;
  avg_packing_time: number; // minutes
  accuracy_rate: number; // percentage

  created_at: Date;
  updated_at: Date;
}

export interface PackingJob {
  id: string;
  sale_id: string;
  sale_number: string;
  picking_list_id: string;
  station_id: string;

  // Customer info
  customer_name: string;
  shipping_address: Address;

  // Packaging
  selected_box?: BoxType;
  weight: number; // kg
  dimensions: {
    length: number;
    width: number;
    height: number;
  };

  // Items
  items: PackingItem[];

  // Status
  status: PackingStatus;
  problem_description?: string;

  // Labels
  shipping_label_url?: string;
  invoice_url?: string;
  packing_slip_url?: string;

  // Responsible
  packed_by?: string;
  packed_by_name?: string;
  packed_at?: Date;

  // Quality check
  quality_checked: boolean;
  quality_checked_by?: string;
  quality_checked_at?: Date;

  created_at: Date;
  updated_at: Date;
}

export interface PackingItem {
  product_id: string;
  product_name: string;
  quantity: number;
  packed: boolean;

  // Verification
  verified: boolean;
  verified_at?: Date;
}

export interface BoxType {
  id: string;
  name: string;
  code: string;

  // Dimensions (cm)
  internal_dimensions: {
    length: number;
    width: number;
    height: number;
  };

  // Limits
  max_weight: number; // kg

  // Cost
  cost: number;

  // Availability
  stock_quantity: number;
  min_stock: number;

  // Usage
  is_active: boolean;
  usage_count: number;

  created_at: Date;
}

// ============================================================================
// SHIPPING & DELIVERY
// ============================================================================

export type DeliveryStatus = 'pending' | 'in_route' | 'delivered' | 'failed' | 'returned';
export type RouteStatus = 'planned' | 'in_progress' | 'completed' | 'cancelled';
export type VehicleStatus = 'available' | 'in_use' | 'maintenance';
export type DeliveryPriority = 'normal' | 'urgent' | 'express';

export interface DeliveryRoute {
  id: string;
  route_number: string;
  date: Date;
  workspace_id: number;

  // Vehicle
  vehicle_id: string;
  vehicle_plate: string;
  driver_id: string;
  driver_name: string;

  // Deliveries
  deliveries: Delivery[];
  optimized_sequence: string[]; // IDs in optimized order

  // Metrics
  total_distance: number; // km
  estimated_time: number; // minutes
  actual_time?: number;
  total_weight: number; // kg
  total_volume: number; // m³
  total_stops: number;

  // Status
  status: RouteStatus;
  started_at?: Date;
  completed_at?: Date;

  // Performance
  on_time_deliveries: number;
  failed_deliveries: number;
  success_rate: number; // percentage

  created_at: Date;
  updated_at: Date;
}

export interface Delivery {
  id: string;
  sale_id: string;
  sale_number: string;
  route_id?: string;
  sequence_number?: number;

  // Recipient
  customer_id: string;
  customer_name: string;
  customer_phone: string;
  address: Address;

  // Coordinates
  latitude?: number;
  longitude?: number;

  // Delivery window
  delivery_window_start?: Date;
  delivery_window_end?: Date;

  // Priority
  priority: DeliveryPriority;

  // Items
  packages: Package[];

  // Status
  status: DeliveryStatus;

  // Delivery
  delivered_at?: Date;
  delivered_by?: string;
  recipient_name?: string;
  signature_url?: string;
  photo_url?: string;

  // Failure
  failed_reason?: string;
  failed_attempts: number;
  next_attempt_date?: Date;

  // Notes
  delivery_notes?: string;
  driver_notes?: string;
  customer_notes?: string;

  // Estimated time at location
  estimated_duration: number; // minutes

  created_at: Date;
  updated_at: Date;
}

export interface Package {
  id: string;
  tracking_code: string;
  weight: number; // kg
  dimensions: {
    length: number;
    width: number;
    height: number;
  };

  // Contents
  items: {
    product_id: string;
    product_name: string;
    quantity: number;
  }[];

  // Handling
  fragile: boolean;
  requires_refrigeration: boolean;
  special_handling?: string;
}

export interface Address {
  street: string;
  number: string;
  complement?: string;
  neighborhood: string;
  city: string;
  state: string;
  postal_code: string;
  country: string;

  // Location
  latitude?: number;
  longitude?: number;

  // References
  reference?: string;
  delivery_instructions?: string;
}

export interface Vehicle {
  id: string;
  license_plate: string;
  model: string;
  brand: string;
  year: number;

  // Capacity
  max_weight: number; // kg
  max_volume: number; // m³

  // Status
  status: VehicleStatus;
  maintenance_due_date?: Date;

  // GPS
  current_location?: {
    latitude: number;
    longitude: number;
    updated_at: Date;
  };

  // Costs
  fuel_consumption: number; // km/l
  daily_cost: number;

  // Odometer
  odometer_km: number;

  created_at: Date;
  updated_at: Date;
}

export interface RouteOptimization {
  deliveries: Delivery[];
  vehicles: Vehicle[];

  // Settings
  optimize_by: 'distance' | 'time' | 'cost';
  consider_traffic: boolean;
  consider_delivery_windows: boolean;
  max_stops_per_route: number;
  start_location?: {
    latitude: number;
    longitude: number;
  };
  end_location?: {
    latitude: number;
    longitude: number;
  };

  // Result
  optimized_routes: DeliveryRoute[];
  total_distance_saved: number;
  total_time_saved: number;
  unassigned_deliveries: string[]; // IDs not allocated
}

// ============================================================================
// SHIPPING CARRIERS
// ============================================================================

export type CarrierCode = 'correios' | 'jadlog' | 'melhor_envio' | 'own_fleet' | 'other';

export interface ShippingCarrier {
  id: string;
  name: string;
  code: CarrierCode;
  workspace_id: number;

  // Credentials
  api_key?: string;
  api_secret?: string;
  api_token?: string;
  api_url?: string;

  // Configuration
  is_active: boolean;
  is_default: boolean;

  // Available services
  services: ShippingService[];

  // Integration
  supports_tracking: boolean;
  supports_label_generation: boolean;
  supports_pickup_scheduling: boolean;
  supports_real_time_quotes: boolean;

  // Settings
  default_service_code?: string;
  auto_generate_labels: boolean;
  auto_send_tracking: boolean;

  created_at: Date;
  updated_at: Date;
}

export interface ShippingService {
  code: string;
  name: string;
  delivery_time: string; // Ex: "2-3 dias úteis"
  is_active: boolean;

  // Restrictions
  max_weight?: number;
  max_dimensions?: {
    length: number;
    width: number;
    height: number;
  };
}

export interface ShippingQuote {
  carrier_id: string;
  carrier_name: string;
  service_code: string;
  service_name: string;

  // Price
  price: number;
  discount: number;
  final_price: number;

  // Deadline
  delivery_time_days: number;
  delivery_date: Date;

  // Dimensions
  calculated_weight: number; // Peso cubado
  actual_weight: number;

  // Availability
  available: boolean;
  error_message?: string;

  // Additional info
  insurance_value?: number;
  pickup_required: boolean;
}

export interface ShipmentTracking {
  tracking_code: string;
  carrier_id: string;
  carrier_name: string;
  sale_id: string;

  // Current status
  current_status: string;
  current_location: string;
  last_update: Date;

  // History
  events: TrackingEvent[];

  // Forecast
  estimated_delivery: Date;
  actual_delivery?: Date;

  // Status classification
  is_delivered: boolean;
  is_delayed: boolean;
  has_issue: boolean;

  created_at: Date;
  updated_at: Date;
}

export interface TrackingEvent {
  date: Date;
  status: string;
  location: string;
  description: string;
  details?: string;
}

// ============================================================================
// LOGISTICS DASHBOARD & ANALYTICS
// ============================================================================

export interface LogisticsDashboard {
  date: Date;
  workspace_id: number;

  // Picking metrics
  picking: {
    total_lists: number;
    completed_lists: number;
    pending_lists: number;
    avg_picking_time: number;
    accuracy_rate: number;
  };

  // Packing metrics
  packing: {
    total_jobs: number;
    completed_jobs: number;
    pending_jobs: number;
    avg_packing_time: number;
    quality_issues: number;
  };

  // Delivery metrics
  delivery: {
    total_deliveries: number;
    completed_deliveries: number;
    in_route_deliveries: number;
    failed_deliveries: number;
    success_rate: number;
    on_time_rate: number;
  };

  // Route metrics
  routes: {
    total_routes: number;
    active_routes: number;
    total_distance_km: number;
    avg_stops_per_route: number;
    optimization_savings: number;
  };

  // Productivity
  productivity: {
    orders_per_hour: number;
    items_picked_per_hour: number;
    packages_packed_per_hour: number;
  };
}

export interface WarehousePerformance {
  warehouse_id: string;
  warehouse_name: string;
  period_start: Date;
  period_end: Date;

  // Throughput
  orders_processed: number;
  items_picked: number;
  packages_shipped: number;

  // Speed
  avg_order_processing_time: number; // minutes
  avg_picking_time: number;
  avg_packing_time: number;

  // Quality
  picking_accuracy: number; // percentage
  packing_accuracy: number;
  damage_rate: number;

  // Efficiency
  space_utilization: number; // percentage
  labor_productivity: number;
  cost_per_order: number;
}

export interface DeliveryPerformance {
  driver_id: string;
  driver_name: string;
  period_start: Date;
  period_end: Date;

  // Volume
  total_deliveries: number;
  successful_deliveries: number;
  failed_deliveries: number;

  // Performance
  on_time_rate: number; // percentage
  success_rate: number;
  avg_deliveries_per_day: number;

  // Efficiency
  total_distance_km: number;
  fuel_efficiency: number; // km/l
  cost_per_delivery: number;

  // Customer satisfaction
  avg_rating?: number;
  complaints: number;
}

File: backend/app/models/logistics.py
"""
Logistics Models
Picking, Packing, Shipping and Route Management
"""
import enum
from datetime import datetime
from sqlalchemy import (
    Column, Integer, String, Float, Boolean, DateTime,
    ForeignKey, Text, JSON, Enum as SQLEnum, Index
)
from sqlalchemy.orm import relationship
from app.db.base_class import Base


# ============================================================================
# ENUMS
# ============================================================================

class PickingType(str, enum.Enum):
    SINGLE_ORDER = "single_order"
    BATCH = "batch"
    WAVE = "wave"


class PickingStatus(str, enum.Enum):
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class PackingStatus(str, enum.Enum):
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    PROBLEM = "problem"


class DeliveryStatus(str, enum.Enum):
    PENDING = "pending"
    IN_ROUTE = "in_route"
    DELIVERED = "delivered"
    FAILED = "failed"
    RETURNED = "returned"


class RouteStatus(str, enum.Enum):
    PLANNED = "planned"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class VehicleStatus(str, enum.Enum):
    AVAILABLE = "available"
    IN_USE = "in_use"
    MAINTENANCE = "maintenance"


class DeliveryPriority(str, enum.Enum):
    NORMAL = "normal"
    URGENT = "urgent"
    EXPRESS = "express"


# ============================================================================
# MODELS
# ============================================================================

class BoxType(Base):
    """Box/Package Types for Packing"""
    __tablename__ = "box_types"

    id = Column(Integer, primary_key=True, index=True)
    workspace_id = Column(Integer, ForeignKey("workspaces.id"), nullable=False, index=True)

    name = Column(String(255), nullable=False)
    code = Column(String(50), nullable=False)

    # Dimensions in cm
    internal_length = Column(Float, nullable=False)
    internal_width = Column(Float, nullable=False)
    internal_height = Column(Float, nullable=False)

    # Limits
    max_weight = Column(Float, nullable=False)  # kg

    # Cost
    cost = Column(Float, nullable=False)

    # Availability
    stock_quantity = Column(Integer, default=0, nullable=False)
    min_stock = Column(Integer, default=0, nullable=False)

    # Usage
    is_active = Column(Boolean, default=True, nullable=False)
    usage_count = Column(Integer, default=0, nullable=False)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    workspace = relationship("Workspace", back_populates="box_types")

    __table_args__ = (
        Index('idx_box_type_workspace', 'workspace_id', 'code'),
    )


class PickingList(Base):
    """Picking Lists for Order Fulfillment"""
    __tablename__ = "picking_lists"

    id = Column(Integer, primary_key=True, index=True)
    picking_number = Column(String(100), nullable=False, unique=True, index=True)
    workspace_id = Column(Integer, ForeignKey("workspaces.id"), nullable=False, index=True)

    # Type
    type = Column(SQLEnum(PickingType), nullable=False)

    # Orders (JSON array of sale IDs)
    sale_ids = Column(JSON, nullable=False)

    # Items (JSON array with product details, locations, etc.)
    items = Column(JSON, nullable=False)

    # Optimized route (JSON array of locations)
    picking_route = Column(JSON, nullable=True)

    # Status
    status = Column(SQLEnum(PickingStatus), default=PickingStatus.PENDING, nullable=False, index=True)

    # Team
    assigned_to = Column(Integer, ForeignKey("users.id"), nullable=True)
    started_at = Column(DateTime, nullable=True)
    completed_at = Column(DateTime, nullable=True)

    # Metrics
    estimated_time = Column(Integer, nullable=False)  # minutes
    actual_time = Column(Integer, nullable=True)  # minutes

    # Priority
    priority = Column(String(20), default='normal', nullable=False)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    workspace = relationship("Workspace", back_populates="picking_lists")
    assigned_user = relationship("User", foreign_keys=[assigned_to])
    packing_jobs = relationship("PackingJob", back_populates="picking_list", cascade="all, delete-orphan")

    __table_args__ = (
        Index('idx_picking_workspace_status', 'workspace_id', 'status'),
        Index('idx_picking_created', 'created_at'),
    )


class PackingStation(Base):
    """Packing Stations in Warehouse"""
    __tablename__ = "packing_stations"

    id = Column(Integer, primary_key=True, index=True)
    workspace_id = Column(Integer, ForeignKey("workspaces.id"), nullable=False, index=True)
    warehouse_id = Column(Integer, ForeignKey("warehouses.id"), nullable=True)

    name = Column(String(255), nullable=False)
    code = Column(String(50), nullable=False)

    # Configuration
    packer_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    is_active = Column(Boolean, default=True, nullable=False)

    # Metrics
    orders_packed_today = Column(Integer, default=0, nullable=False)
    avg_packing_time = Column(Float, default=0, nullable=False)  # minutes
    accuracy_rate = Column(Float, default=100, nullable=False)  # percentage

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    workspace = relationship("Workspace", back_populates="packing_stations")
    warehouse = relationship("Warehouse")
    packer = relationship("User", foreign_keys=[packer_id])
    packing_jobs = relationship("PackingJob", back_populates="station")

    __table_args__ = (
        Index('idx_packing_station_workspace', 'workspace_id', 'code'),
    )


class PackingJob(Base):
    """Packing Jobs for Orders"""
    __tablename__ = "packing_jobs"

    id = Column(Integer, primary_key=True, index=True)
    workspace_id = Column(Integer, ForeignKey("workspaces.id"), nullable=False, index=True)
    sale_id = Column(Integer, ForeignKey("sales.id"), nullable=False, index=True)
    picking_list_id = Column(Integer, ForeignKey("picking_lists.id"), nullable=True)
    station_id = Column(Integer, ForeignKey("packing_stations.id"), nullable=True)

    # Customer info (denormalized for performance)
    customer_name = Column(String(255), nullable=False)
    shipping_address = Column(JSON, nullable=False)

    # Packaging
    selected_box_id = Column(Integer, ForeignKey("box_types.id"), nullable=True)
    weight = Column(Float, nullable=False)  # kg
    dimensions = Column(JSON, nullable=False)  # length, width, height

    # Items (JSON array)
    items = Column(JSON, nullable=False)

    # Status
    status = Column(SQLEnum(PackingStatus), default=PackingStatus.PENDING, nullable=False, index=True)
    problem_description = Column(Text, nullable=True)

    # Labels and documents
    shipping_label_url = Column(Text, nullable=True)
    invoice_url = Column(Text, nullable=True)
    packing_slip_url = Column(Text, nullable=True)

    # Responsible
    packed_by = Column(Integer, ForeignKey("users.id"), nullable=True)
    packed_at = Column(DateTime, nullable=True)

    # Quality check
    quality_checked = Column(Boolean, default=False, nullable=False)
    quality_checked_by = Column(Integer, ForeignKey("users.id"), nullable=True)
    quality_checked_at = Column(DateTime, nullable=True)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    workspace = relationship("Workspace", back_populates="packing_jobs")
    sale = relationship("Sale")
    picking_list = relationship("PickingList", back_populates="packing_jobs")
    station = relationship("PackingStation", back_populates="packing_jobs")
    selected_box = relationship("BoxType")
    packer = relationship("User", foreign_keys=[packed_by])
    quality_checker = relationship("User", foreign_keys=[quality_checked_by])

    __table_args__ = (
        Index('idx_packing_job_workspace_status', 'workspace_id', 'status'),
        Index('idx_packing_job_sale', 'sale_id'),
    )


class Vehicle(Base):
    """Delivery Vehicles"""
    __tablename__ = "vehicles"

    id = Column(Integer, primary_key=True, index=True)
    workspace_id = Column(Integer, ForeignKey("workspaces.id"), nullable=False, index=True)

    license_plate = Column(String(20), nullable=False, unique=True, index=True)
    model = Column(String(100), nullable=False)
    brand = Column(String(100), nullable=False)
    year = Column(Integer, nullable=False)

    # Capacity
    max_weight = Column(Float, nullable=False)  # kg
    max_volume = Column(Float, nullable=False)  # m³

    # Status
    status = Column(SQLEnum(VehicleStatus), default=VehicleStatus.AVAILABLE, nullable=False, index=True)

    # Current location (JSON: latitude, longitude, updated_at)
    current_location = Column(JSON, nullable=True)

    # Costs
    fuel_consumption = Column(Float, nullable=True)  # km/l
    daily_cost = Column(Float, nullable=True)

    # Maintenance
    odometer_km = Column(Float, default=0, nullable=False)
    last_maintenance_at = Column(DateTime, nullable=True)
    next_maintenance_km = Column(Float, nullable=True)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    workspace = relationship("Workspace", back_populates="vehicles")
    delivery_routes = relationship("DeliveryRoute", back_populates="vehicle")

    __table_args__ = (
        Index('idx_vehicle_workspace_status', 'workspace_id', 'status'),
    )


class DeliveryRoute(Base):
    """Delivery Routes"""
    __tablename__ = "delivery_routes"

    id = Column(Integer, primary_key=True, index=True)
    route_number = Column(String(100), nullable=False, unique=True, index=True)
    workspace_id = Column(Integer, ForeignKey("workspaces.id"), nullable=False, index=True)
    date = Column(DateTime, nullable=False, index=True)

    # Vehicle
    vehicle_id = Column(Integer, ForeignKey("vehicles.id"), nullable=False, index=True)

    # Driver
    driver_id = Column(Integer, ForeignKey("users.id"), nullable=True)

    # Route data
    total_distance_km = Column(Float, default=0, nullable=False)
    estimated_duration = Column(Integer, default=0, nullable=False)  # minutes
    actual_duration = Column(Integer, nullable=True)  # minutes

    # Optimization
    optimized = Column(Boolean, default=False, nullable=False)
    optimization_savings = Column(Float, nullable=True)  # percentage

    # Status
    status = Column(SQLEnum(RouteStatus), default=RouteStatus.PLANNED, nullable=False, index=True)

    # Timing
    started_at = Column(DateTime, nullable=True)
    completed_at = Column(DateTime, nullable=True)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    workspace = relationship("Workspace", back_populates="delivery_routes")
    vehicle = relationship("Vehicle", back_populates="delivery_routes")
    driver = relationship("User", foreign_keys=[driver_id])
    deliveries = relationship("Delivery", back_populates="route", cascade="all, delete-orphan")

    __table_args__ = (
        Index('idx_delivery_route_workspace_date', 'workspace_id', 'date'),
        Index('idx_delivery_route_status', 'status'),
    )


class Delivery(Base):
    """Individual Deliveries"""
    __tablename__ = "deliveries"

    id = Column(Integer, primary_key=True, index=True)
    workspace_id = Column(Integer, ForeignKey("workspaces.id"), nullable=False, index=True)
    sale_id = Column(Integer, ForeignKey("sales.id"), nullable=False, index=True)
    route_id = Column(Integer, ForeignKey("delivery_routes.id"), nullable=True, index=True)

    # Sequence in route
    sequence_number = Column(Integer, nullable=True)

    # Customer info
    customer_id = Column(Integer, ForeignKey("customers.id"), nullable=True)
    customer_name = Column(String(255), nullable=False)
    customer_phone = Column(String(50), nullable=True)

    # Address
    address = Column(JSON, nullable=False)
    latitude = Column(Float, nullable=True)
    longitude = Column(Float, nullable=True)

    # Delivery window
    delivery_window_start = Column(DateTime, nullable=True)
    delivery_window_end = Column(DateTime, nullable=True)

    # Priority
    priority = Column(SQLEnum(DeliveryPriority), default=DeliveryPriority.NORMAL, nullable=False)

    # Packages (JSON array)
    packages = Column(JSON, nullable=False)

    # Status
    status = Column(SQLEnum(DeliveryStatus), default=DeliveryStatus.PENDING, nullable=False, index=True)

    # Delivery details
    delivered_at = Column(DateTime, nullable=True)
    delivered_by = Column(Integer, ForeignKey("users.id"), nullable=True)
    proof_of_delivery_url = Column(Text, nullable=True)
    recipient_name = Column(String(255), nullable=True)
    recipient_document = Column(String(50), nullable=True)
    signature_url = Column(Text, nullable=True)

    # Failed delivery
    failed_at = Column(DateTime, nullable=True)
    failure_reason = Column(Text, nullable=True)
    reschedule_date = Column(DateTime, nullable=True)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    workspace = relationship("Workspace", back_populates="deliveries")
    sale = relationship("Sale")
    route = relationship("DeliveryRoute", back_populates="deliveries")
    customer = relationship("Customer")
    delivered_by_user = relationship("User", foreign_keys=[delivered_by])

    __table_args__ = (
        Index('idx_delivery_workspace_status', 'workspace_id', 'status'),
        Index('idx_delivery_route_sequence', 'route_id', 'sequence_number'),
        Index('idx_delivery_window', 'delivery_window_start'),
    )
